pkg/apputils/utils.go

```diff
@@ -0,0 +1,191 @@
+package apputils
+
+import (
+	"context"
+	"fmt"
+
+	wego "github.com/weaveworks/weave-gitops/api/v1alpha1"
+	"github.com/weaveworks/weave-gitops/pkg/flux"
+	"github.com/weaveworks/weave-gitops/pkg/git"
+	"github.com/weaveworks/weave-gitops/pkg/gitproviders"
+	"github.com/weaveworks/weave-gitops/pkg/kube"
+	"github.com/weaveworks/weave-gitops/pkg/logger"
+	"github.com/weaveworks/weave-gitops/pkg/osys"
+	"github.com/weaveworks/weave-gitops/pkg/runner"
+	"github.com/weaveworks/weave-gitops/pkg/services/app"
+	"github.com/weaveworks/weave-gitops/pkg/services/auth"
+	"github.com/weaveworks/weave-gitops/pkg/utils"
+	"k8s.io/apimachinery/pkg/types"
+)
+
+//go:generate go run github.com/maxbrunsfeld/counterfeiter/v6 -generate
+//counterfeiter:generate . AppFactory
+type AppFactory interface {
+	GetKubeService() (kube.Kube, error)
+	GetAppService(ctx context.Context, name, namespace string) (*app.App, error)
+}
+
+type DefaultAppFactory struct {
+}
+
+func (f *DefaultAppFactory) GetAppService(ctx context.Context, name, namespace string) (*app.App, error) {
+	return GetAppService(ctx, name, namespace)
+}
+
+func (f *DefaultAppFactory) GetKubeService() (kube.Kube, error) {
+	kubeClient, _, kubeErr := kube.NewKubeHTTPClient()
```

Comment:
Not sure what the implications of multiple k8s clients will be. Each will make a separate TCP connection and teardown when its done (I assume). It might already do that with the singleton client.

In any case, something to keep an eye on.

Reply:
Sounds good. We also do this in the paths with an app service so it will require some more work if it turns out to be a significant bottleneck.